File: python/tests/helpers/mock_relay.py
from __future__ import annotations

import json
from typing import Any

import websockets


class MockRelay:
    """In-process WebSocket server on a random port, mirroring
    typescript/tests/helpers/mock-relay.ts. Auto-replies `auth_ok` to `auth`."""

    def __init__(self) -> None:
        self.received: list[dict[str, Any]] = []
        self._active_ws: Any = None
        self._server: Any = None

    @property
    def port(self) -> int:
        assert self._server is not None
        return self._server.sockets[0].getsockname()[1]

    async def _handle(self, ws: Any) -> None:
        self._active_ws = ws
        try:
            async for raw in ws:
                msg = json.loads(raw)
                self.received.append(msg)
                if msg.get("type") == "auth":
                    await ws.send(
                        json.dumps({"type": "auth_ok", "agentId": "mock-agent-id", "relayInstanceId": "mock-relay"})
                    )
        finally:
            if self._active_ws is ws:
                self._active_ws = None

    async def start(self) -> None:
        self._server = await websockets.serve(self._handle, "localhost", 0)

    async def dispatch(self, call_id: str, input: Any, schema_version: int = 1) -> None:
        """Dispatch a call to the agent (after auth)."""
        if self._active_ws is None:
            raise RuntimeError("No active WebSocket connection to dispatch to")
        await self._active_ws.send(
            json.dumps({"type": "call", "callId": call_id, "schemaVersion": schema_version, "input": input})
        )

    async def close_connections(self) -> None:
        """Force-close the active connected client WebSocket (triggers agent reconnect)."""
        if self._active_ws is not None:
            await self._active_ws.close()

    async def close(self) -> None:
        assert self._server is not None
        self._server.close()
        await self._server.wait_closed()


async def create_mock_relay() -> MockRelay:
    relay = MockRelay()
    await relay.start()
    return relay

File: typescript/tests/helpers/mock-relay.ts
import { WebSocketServer, WebSocket } from 'ws'
import type { AddressInfo } from 'net'

export interface MockRelay {
  port: number
  /** All messages received from any connected agent */
  received: unknown[]
  /** Dispatch a call to the agent (after auth) */
  dispatch(callId: string, input: unknown, schemaVersion?: number): void
  /** Force-close all connected client WebSockets (triggers agent reconnect) */
  closeConnections(): void
  /** Close the server */
  close(): Promise<void>
}

export function createMockRelay(): MockRelay {
  const wss = new WebSocketServer({ port: 0 })
  const port = (wss.address() as AddressInfo).port
  const received: unknown[] = []

  let activeWs: WebSocket | null = null

  wss.on('connection', (ws) => {
    activeWs = ws

    ws.on('message', (raw) => {
      const msg = JSON.parse(raw.toString())
      received.push(msg)

      if (msg.type === 'auth') {
        ws.send(
          JSON.stringify({ type: 'auth_ok', agentId: 'mock-agent-id', relayInstanceId: 'mock-relay' }),
        )
      }
    })

    ws.on('close', () => {
      if (activeWs === ws) activeWs = null
    })
  })

  return {
    port,
    received,

    closeConnections() {
      for (const client of wss.clients) {
        client.terminate()
      }
    },

    dispatch(callId, input, schemaVersion = 1) {
      if (!activeWs || activeWs.readyState !== WebSocket.OPEN) {
        throw new Error('No active WebSocket connection to dispatch to')
      }
      activeWs.send(JSON.stringify({ type: 'call', callId, schemaVersion, input }))
    },

    close() {
      return new Promise((res, rej) => wss.close((err) => (err ? rej(err) : res())))
    },
  }
}
